[1-4 단계 - 계산기 구현] 천희정 과제 제출합니다.

heejung72 opened the pull request:
안녕하세요 오찌님 !
계산기 1-4단계 미션 제출합니다.

이번 첫주차 스터디를 진행하며 자바에 대한 이해가 좀 더 명확히 정리되어야겠다는 점을 느꼈습니다. 간단해 보이는 문제라도 풀어보니 예상보다 시간이 걸렸고, 이를 통해 더 많은 공부가 필요하다는 것을 깨달았습니다. 두 달동안 주시는 피드백을 최대한 체화하며 실력을 쌓아가기위해 노력하겠습니다. 피드백 주실 부분이 많을 것 같아서 몇가지 궁금한 점을 추가로 적어봅니다.

1. 단일 책임 원칙 (SRP) 적용
현재 코드에서 구분자를 기준으로 분리, 예외처리, 계산 로직이 한 메서드에 포함되어 있습니다. 이는 이번 미션에서는 로직이 엄청 복잡하지는 않아서 한번에 포함해 작성했습니다. 그럼에도 이를 분리하는게 더 좋은지 아니면 지금처럼 하나로 처리하는게 더 좋은지가 궁금합니다. 

2. 구조 확장성
현재의 구조가 다른 연산자나 사용자 정의 구분자를 처리하는데 확장성이 있을지 궁금합니다. 새로운 기능을 추가할 때 현재 구조에서 쉽게 확장이 가능한지, 아니면 다른 구조로 바꾸는게 더 좋다면 어떤식으로 바꿔야하는지 궁금합니다.

3. 입력 형식 처리
처음에는 String[]으로 입력값을 받았는데, 문자열 처리가 어려운 점이 많았습니다. 구체적으로 어떤 방식으로 입력을 처리하는게 더 유연하고 오류를 줄일 지 궁금합니다.

4. 테스트 케이스
현재 테스트 케이스에서 음수, 숫자가 아닌 것, 여러 구분자 혼합 등을 잘 다루었는지 궁금합니다. 이외에도 추가로 넣어야 할 테스트 케이스가 있을지 궁금합니다.

5. 코드 개선
추가로 현재 코드에서 리팩토링할 부분이나, 성능, 가독성, 유지보수 측면에서 개선해야할 점이 있는지 궁금합니다.

마지막으로, PR방식이나 commit한 방식이 제대로 된 방식인지도 궁금합니다.
피드백을 주시면 최대한 반영해 개선하겠습니다.
감사합니다 !

heejung72 committed:
feat : 1단계 초간단 계산기 구현

- 인자 2개를 받아 사칙연산을 할 수 있는 계산기를 구현한다.
- 사칙연산과 매칭되는 4개의 메서드를 제공한다.
- 계산된 결과는 정수를 반환한다.
- 메인 메서드는 만들지 않는다.

heejung72 committed:
feat : 2단계 초간단 문자열 테스트

- 구현한 초간단 계산기가 예상한대로 동작하는지 Junit5를 활용하여 테스트를 자동화한다.
- 메인 메서드 없이 동작을 검증하는 경험을 해본다.
- 메인 메서드는 만들지 않는다.

heejung72 committed:
feat : 3단계 문자열 계산기 구현

- 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환 (예: “” => 0, "1,2" => 3, "1,2,3" => 6, “1,2:3” => 6)
- 앞의 기본 구분자(쉼표, 콜론)외에 커스텀 구분자를 지정할 수 있다. 커스텀 구분자는 문자열 앞부분의 “//”와 “\n” 사이에 위치하는 문자를 커스텀 구분자로 사용한다. 예를 들어 “//;\n1;2;3”과 같이 값을 입력할 경우 커스텀 구분자는 세미콜론(;)이며, 결과 값은 6이 반환되어야 한다.
- 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw한다.
- 구현한 문자열 계산기가 예상한대로 동작하는지 Junit5를 활용하여 테스트를 자동화한다.
- 조금 더 복잡한 도메인을 대상으로 테스트를 작성하는 경험을 해본다.
- 메인 메서드는 만들지 않는다.

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -0,0 +1,52 @@
+    public int add(String input) {
+        if (input == null || input.trim().isEmpty()) {
+            return 0;
+        }
```
null 체크 뿐 아니라 공백만 있는 케이스도 체크해주셨군요! trim().isEmpty()같은 경우에는 자바 11 이상부터는 isBlank() 라는 메서드가 있기 때문에 해당 메서드를 사용해보셔도 좋을 것 같아요!

heejung72 replied:
isEmpty() 대신 isBlank() 사용해보며 코드를 변경했습니다 !

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -0,0 +1,52 @@
+            String[] parts = input.split("\n", 2);
+            delimiter = Pattern.quote(parts[0].substring(2));
+            numbers = parts[1];
+        }
```
몇 가지 리뷰를 드릴 수 있을 것 같아요.

1. delimiter 라는 이름으로 ",|:" 라는 문자열을 통으로 넘겨주고 있는데, `,":` 가 통째로 구분자가 되는 것이 아니라 각각이 구분자이기 때문에 delimiter라는 네이밍은 오해의 소지가 있지 않을까요?
2. delimiter에 할당해주는 문자열 리터럴이 `|`를 포함하고 있는 이유는 무엇일까요? 요구사항에 존재하는 기본 구분자는 콤마와 콜론 뿐입니다!
3. delimiter가 아래에서 재정의 되고 있네요. 이렇게 변수를 재정의 하는 경우 어떤 문제가 발생할 수 있을까요?

heejung72 replied:
1. 생각해보니 명확하지 않은 변수명을 작성한 것 같아, String defaultDelimiterRegularExpression으로 바꾸었습니다 
2. | 이 OR의 의미를 가지고 있어서 코드를 저렇게 작성하였는데, 정규표현식을 사용해 String defaultDelimiterRegularExpression = "[,:]"; 로 수정하였습니다 .
3. delimiter를 재정의 할 경우, 코드의 가독성이 떨어지고, 의미 전달에 오류가 발생할 수 있습니다. 따라서, defaultDelimiterRegularExpression, delimiterRegularExpression으로 변수를 나누었습니다.

코드를 수정하면서 가장 크게 궁금한 점은 변수명을 보다 명확하게 바꾸기위해 노력했는데, 변수명이 너무 길어서 오히려 좋지 않은 이름을 붙인건가 싶기도 합니다 ...

Ohzzi replied:
저는 개인적으로는 길더라도 명확한 네이밍을 더 선호하는 편인데요! 너무 길다 싶으면 축약어를 쓰거나 암묵적으로 알고 넘어갈만한 요소는 생략해도 좋습니다. 가령, delimiterRegularExpression이 너무 길다면
- delimiterExpression (Regular 생략)

혹은
- delimiterRegex (RegularExpression의 축약어. 일반적으로 regex 라고 많이 사용하는 편)

이런 식으로 사용할 수 있을 것 같아요!

heejung72 replied:
확인했습니다 ! 좋은 예시 감사합니다

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -0,0 +1,52 @@
+            try {
+                int number = Integer.parseInt(token.trim());
+                if (number < 0) {
+                    if (negatives.length() > 0) {
```
optional) 요건 isEmpty의 부정형으로 대체할 수 있을 것 같아요!

heejung72 replied:
negatives.length() > 0 은 !negatives.isEmpty()로 대체하였습니다 !

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -0,0 +1,52 @@
+                    if (negatives.length() > 0) {
+                        negatives.append(", ");
+                    }
+                    negatives.append(number);
```
number가 0 미만이면 바로 예외를 throw 해도 되지 않을까요? 현재의 negatives StringBuilder 사용으로도 물론 결과를 도출할 수 있지만 불필요한 동작이 늘어난 것 같아요

```suggestion
                if (number < 0) {
                    throw new RuntimeException("음수는 안 됩니다. 입력된 값: " + negatives);
                }
```

heejung72 replied:
어떠한 음수 값을 입력했는지 보여주고 싶어서 저 코드를 작성하였는데, 의미가 없을 것 같기도 합니다. 따라서, 말씀해주신 것 처럼 바로 예외처리를 진행하는 방향으로 수정하였습니다.

Ohzzi commented on src/main/java/NumberCalc/Input.java:
```diff
@@ -0,0 +1,10 @@
+
+    int n1 = sc.nextInt();
+    int n2 = sc.nextInt();
+}
```
(input은 미션 요구사항이 아니지만) main 함수가 있는 프로그램을 작성하고, input도 프로그램 안에서 여러번 받아야 한다고 가정해볼게요. 그렇다면 input을 받아야 할 때마다 Input 클래스를 생성해야 하고 Scanner도 생성해야 할텐데요, 문제가 되는 부분은 없을까요?

heejung72 replied:
Scanner 의 중복 생성에 어떠한 문제가 있는지 찾아보았습니다.
   

>  Input input1 = new Input(); // 내부적으로 Scanner 하나 생성
>     Input input2 = new Input(); // 또 Scanner 하나 생성

이 경우 Scanner 객체가 System.in을 여러번 참조하게 됩니다.
하지만, System.in은 표준 입력 시스템 하나만 존재합니다.
따라서, 여러개의 스캐너가 이 스트림을 동시에 사용하려고하면 1. 입력 충돌 
2. NoSuchElementException 
3. IllegalStateException 
4. 하나의 Scanner가 입력을 닫을 경우 다른 스캐너도 입력 못받는 문제
총 4가지의 문제가 생길 수 있습니다.
이것의 해결방안으로는 스캐너를 한번만 생성해 공유하거나, Scanner를 메인에서 직접 관리하면 됩니다.

저는 스캐너를 한번만 생성해 공유하는 방식으로 리팩토링 하였습니다 !

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -0,0 +1,52 @@
+
+        if (input.startsWith("//")) {
+            String[] parts = input.split("\n", 2);
+            delimiter = Pattern.quote(parts[0].substring(2));
```
quote 처리를 안하게 되면 어떻게 되나요?

heejung72 replied:
quote 는 정규 표현식의 특수문자들을 일반 문자로 처리하기 위해 사용했습니다. 이를 생략하면 사용자가 지정한 구분자에 정규식에서 특별한 의미를 가지는 문자가 포함되어있을 경우 문제가 발생합니다.
예를들면 
String input = "//|\n1|2|3";
String delimiter = "|";
여기서,
String[] tokens = "1|2|3".split(delimiter);일 경우 
["", "1", "", "2", "", "3", ""]이렇게 결과가 나옵니다.
이는 |가 정규식에서 OR의 의미를 가지기 때문에 그렇습니다.

Ohzzi commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+        @DisplayName("add는 x, y를 입력받아 덧셈 결과를 반환한다")
+        void addTest(int x, int y, int expected) {
+            int result = calc.add(x, y);
+            assertEquals(expected, result);
```
> 기존 JUnit5로 작성되어 있던 단위 테스트를 AssertJ로 리팩터링한다.
JUnit5에서 제공하는 기능과 AssertJ에서 제공하는 기능을 사용해보고, 어떠한 차이가 있는지 경험한다.

4단계 요구사항이 테스트 리팩터링인데요! assertJ의 것을 사용하고 있지 않은 것 같아요!

heejung72 replied:
AssertJ를 사용해서 테스트 리팩토링 진행한 부분 step4 브랜치에 올렸습니다 !

Ohzzi replied:
아하! 지금 step3 브랜치가 PR의 머지 대상으로 되어 있어요! cherry-pick 등 기능을 사용해서 step4 요구사항도 그냥 step3로 반영해주시면 좋을 것 같습니다!

heejung72 replied:
step4에 작성한 내용 step3로 merge 하였습니다 !

Ohzzi commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+    @DisplayName("사칙연산 테스트")
+    class OperationTest {
+
+        @ParameterizedTest
```
DisplayName으로 이 ParameterizedTest에 통째로 네이밍을 지정할 수 있지만 `@ParameterizedTest` 어노테이션의 파라미터를 통해 각각의 테스트 케이스에 대해서도 네이밍을 정할 수 있습니다! 한 번 살펴보셔도 좋을 것 같아요!

heejung72 replied:
네이밍 하는 방법 중 DisplayName만 사용했는데, DisplayName 말고 ParameterizedTest(name = “…”) 를 사용해 이름을 줄 수 있음을 배웠습니다 !
따라서, 

> @ParameterizedTest(name = "1 + 2 = 3")

> // @MethodSource("addArguments")

이렇게 수정하여 사용해보았습니다

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -0,0 +1,52 @@
+        StringBuilder negatives = new StringBuilder();
+
+        for (String token : tokens) {
+            if (token.trim().isEmpty()) {
+                continue;
+            }
```
delmiter로 쪼갠 문자열 토큰이 빈 값이라면 구분자가 연속으로 2번 이상 나왔다는 것인데 예외상황이 아닐까요?

heejung72 replied:
구분자가 연속으로 2번 이상 나온 경우에는 그냥 공백이라고 생각하여 continue를 사용하였는데, 예외 처리로 수정하였습니다 !
> if (token.trim().isEmpty()) {
>                 throw new RuntimeException("구분자 사이에 값이 없습니다.");
>             }

Ohzzi commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+            assertEquals(expected, result);
+        }
+
+        @ParameterizedTest
+        @MethodSource("subtractArguments")
+        @DisplayName("subtract는 x, y를 입력받아 뺄셈 결과를 반환한다")
```
`@MethodSource`는 여러 가지 테스트 케이스에 대해 복잡한 오브젝트를 전달하기에 좋은 테스트 방식인데요, 대신 테스트 메서드 하나만으로 테스트 케이스를 파악할 수 없고 argument를 제공하는 메서드로 왔다갔다 하면서 봐야 하기 때문에 가독성이 떨어질 수 있는 테스트 방식이라고 생각해요. 그래서 이렇게 간단한 테스트에 `@MethodSource`가 효율적인지, 더 효율적인 테스트 방식은 없을지 고민해보시면 좋을 것 같습니다!

heejung72 replied:
@MethodSource는 유연하고 강력하여 복잡한 테스트를 진행할 때 사용한다고 공부했습니다. 하지만 단순 입력에 대해 사용하면 과할 수 있다는 의견도 있군요. 간단한 값을 테스트할 때는 @CscSource, @ValueSource를 사용하는게 더 간결하다고하여 @CscSoure를 사용해보았습니다 !

Ohzzi reviewed (changes requested):
희정님 수고하셨습니다! 코드를 깔끔하게 잘 작성해주셨네요 😄 

리뷰에 앞서, EOF 문제부터 체크하고 넘어갈게요! 파일의 마지막에는 반드시 빈 공백이 한 줄 들어가야 하는데, 일부 파일들의 끝에 공백이 없는 것 같아요! Files Changed 상에서 맨 마지막에 (-) 표시가 있는 파일들이 그렇습니다. 인텔리제이에는 파일에 자동으로 공백 한 줄을 넣어주는 옵션이 있는데 해당 옵션을 찾아보고 적용해보시면 좋을 것 같아요!

아래에는 남겨주신 질문사항에 대한 답변을 드릴게요!

## 단일 책임 원칙 (SRP) 적용
말씀하신 것처럼 이번 미션은 그렇게 복잡하지 않아서 StringCalculator 하나에 모든 로직을 몰아넣어도 크게 어려움이 없을거에요! 하지만 요구사항이 복잡해질수록, 그리고 추가될수록 책임에 따라 객체를 분리하는 것이 가독성 및 유지보수에 훨씬 좋은 코드가 될거에요. 이번 미션도 리뷰 반영 과정에서 책임에 맞게 일부 객체를 나눠보시는 것은 어떨까요?

## 구조의 확장성
만약 기본 구분자가 `,`와 `|` 에서 다른 문자로 바뀐다면 어떨까요? 지금 코드에서는 StringCalculator 안의 delimiter만 바꿔주면 되겠죠. 이렇게 보면 확장성이 좋다고 볼 수 있지만, 코드가 커진다면 StringCalculator.add 중간에 문자열 리터럴로 선언되어 있는 delimiter를 찾기가 어려울거에요. 때문에 상수 처리 등 적절한 처리를 해주는 것이 코드를 확장해나기 더 좋을 것 같습니다.

추가적으로 이 미션 자체가 확장성을 따질만큼 복잡한 요구사항을 가지지 않기 때문에, 현재 이 요구사항을 가지고 확장성을 가지는 구조를 만들려면 어떻게 해야 하는지 구체적인 방향을 잡기에는 어려움이 있어요. 이런 확장성에 대한 고민은 조금 더 많은 요구사항을 가지는 다른 미션들을 진행하면서 직접 느껴보면서 차차 배워나가면 좋을 것 같습니다!

## 입력 형식 처리
문자열을 처리하면서 어떤 부분이 어려웠나요? 구체적으로 말씀 남겨주시면 더욱 도움이 될 것 같습니다!  input 타입도 여러가지로 바꿔보고, 다양한 파싱 로직을 경험해보면서 효율적이고 가독성이 좋다고 느끼는 파싱 방법을 직접 찾아보는게 중요하다고 생각해요. 결국 이 부분도 정답이 없기 때문에 개인 취향이 작용하는 부분이거든요.

콘솔 프로그램을 만든다면 어쩔 수 없이 String으로 입력을 받고, 적절한 타입으로 파싱하는 과정을 거쳐야 할거에요. 여담으로 사실 단지 콘솔 프로그램 뿐 아니라, 웹 서버 애플리케이션 조차도 네트워크로 넘어오는 바이트를 적절한 객체로 deserialize 하는 과정을 거치거든요. (물론 프레임워크가 대부분 이런 역할을 해주지만요) 그래서 적절한 파싱 방법을 찾는 것도 은근 중요하답니다. 저도 실무에서 바이트로 된 스트림을 저희 서버 상에서 사용할 클래스로 어떻게 하면 효율적으로 파싱할 수 있을까(심지어 이 경우는 input은 상대가 넣어줘서 포맷을 변경할 수 없었어요!) 고민을 많이 해봤던 기억도 나네요.

## 테스트
현재 요구사항에 대한 테스트 케이스는 적절하게 들어있는 것 같아요! 요구사항에서 명시한 예외 발생에 대한 케이스도 있고, 구분자가 기본이냐 커스텀이냐에 따른 테스트도 잘 분기가 되어 있어요. 다만 리뷰 본문에도 남겼지만 모든 테스트가 `@MethodSource`로 이루어져 있는데, `@MethodSource`가 과연 모든 상황에서 효율적인 테스트인지 고민해보면 좋을 것 같아요. 많은 케이스를 테스트하는 것도 좋지만, 테스트 코드가 요구사항에 대한 일종의 문서 역할을 할 수 있기 때문에 테스트의 가독성을 챙기는 고민도 필요하다고 생각합니다!

## 코드 개선
코드 개선 및 리팩터링할 방향은 리뷰 본문에 남겨두었습니다! 조금 큰 변경으로는 PR 본문에 남겨주신 것처럼 적절히 책임을 위임한 객체로 쪼개보는 리팩터링을 진행해보면 좋을 것 같아요!

heejung72 committed:
리팩토링

- EOF
- Delimiter 변수이름, 기본 구분자 다시 설정
- Delimiter 재정의 안하고, 변수 두개 설정
- number이 0 미만이면 예외 바로 throw함
- Scanner 중복 생성 안함 (스캐너 한 번 사용해 공유)

heejung72 committed:
feat : 4단계 리팩터링

- build.gradle에 AssertJ 의존성을 추가한다.
- 기존 JUnit5로 작성되어 있던 단위 테스트를 AssertJ로 리팩터링한다.
- JUnit5에서 제공하는 기능과 AssertJ에서 제공하는 기능을 사용해보고, 어떠한 차이가 있는지 경험한다.
- 메인 메서드는 만들지 않는다.

heejung72 committed:
refactor : CsvSource 사용, 예외처리

- MethodSource -> CsvSource 대체
- 구분자가 연속으로 두 번 나오는 경우는 예외처리 진행

heejung72 committed:
refactor : SRP 적용

- StringCalculator 내의
1. 커스텀 및 기본 구분자를 분리하는 DelimiterParsing
2. 정수 파싱하고 음수인지, 숫자인지 확인하는 NumberParsing
3. 구분자를 기준으로 숫자를 뽑아내는 NumberTokens
으로 나누어 SRP 를 적용함

heejung72 committed:
refactor : 줄바꿈 삭제

Ohzzi commented on src/main/java/NumberCalc/Input.java:
```diff
@@ -3,8 +3,17 @@
 
-    int n1 = sc.nextInt();
-    int n2 = sc.nextInt();
+    public Input(Scanner sc) {
+        this.sc = sc;
+    }
+
+    public int readFirstNumber() {
+        return sc.nextInt();
+    }
+
+    public int readSecondNumber() {
+        return sc.nextInt();
+    }
```
Scanner 한 번 생성해서 공유라고 수정해주셨지만, 사실 Input 클래스를 여러 개 만들게 되면 서로 다른 Scanner를 넣을 수 있겠군요. 때문에 아래와 같은 방향을 제안드려볼게요!
```java
pubic class Input {
    private final Scanner sc = new Scanner(System.in);

    private Input() {} // Input의 인스턴스가 필요 없으므로 인스턴스가 생성되지 않도록 방지

    public static int readFirstNumber() {
        resturn sc.nextInt();
    }

    public int readSecondNumber() {
        return sc.nextInt();
    }
}

뷰는 이번 미션의 구현 대상이 아니니 반영은 안하셔도 됩니다!

Ohzzi commented on src/main/java/StringCalc/StringCalculator.java:
```diff
@@ -4,21 +4,26 @@
     public int add(String input) {
-        if (input == null || input.trim().isEmpty()) {
+        // isEmpty() 대신 isBlank() 사용
+        if (input == null || input.trim().isBlank()) {
```
isBlank를 사용하면 trim은 불필요할 것 같아요!

Ohzzi reviewed (approved):
계산기 미션 진행하느라 수고하셨습니다! 수정해주신 부분에 조금 리뷰사항이 있는데 다음 미션에서 반영해주셔도 충분합니다! 추가적인 질문 있으면 남겨주시고 다음 미션에서 뵐게요!!

heejung72 committed:
refactor : 최종 리뷰 반영

- isBlank() 사용으로 인해 trim 삭제
- Scanner 최종 리뷰 반영
- step4 -> step3 머지
